### ISSUE TITLE
slide 8 map vs flatmap

### ISSUE THREAD

**Issue by arinng5** [nus-cs2030/2324-s2#556]
based on what prof mentioned in the lecture, is it correct to say that if you are using two maps in a row, you ALWAYS have to use flatMap before map to avoid the 'bad return type' error?

**Comment by yushanedi** [comment]
No, not necessarily, imo I use flatMaps if my return type for your mapping function produces an _Optional_ or _Stream_, then the flatMap will be useful to unwrap the values and map it accordingly

**Comment by LimZiXiong** [comment]
Take any wrapper like Optional with an element x wrapped inside. If the function takes in a x and produce Optional(y), you use flatMap here. For example it is possible to use map twice and i still want to map. E.g there is 2 function. One maps x -> y and other maps y -> z. In this case I would use map twice, once to turn x into y so i can then use the 2nd function to map y to z

**Comment by chinwn** [comment]
The lecture note is referring to two generators (or unboxing), not
two maps. Here the two generators are m1 and m2

```
   Log<Pair<A,B>> seqP(Log<A> m1, Log<B> m2) {
    // below is a do-notation syntactic sugar
    do {
	    x <-m1;  // view this generator as an unboxing
	    y <-m2;  // view this 2nd generator as other unboxing
	    unit(new Pair(x,y))
	}
   }
```
  
  The do-comprehension is a syntactic sugar notation for
  flatMap. If you follow the translatio rules given, 
  you will get:
  
```
  Log<Pair<A,B>> seqP(Log<A> m1, Log<B> m2) {
      // below is a first-order syntactic sugar
      m1.flatMap( x ->
	    m2.flatMap (y -> unit(new Pair(x,y)))
  }
```
  
  This is also equivalent to below where the
  last generator is translated to a map instead
  of flatMap.

```
  Log<Pair<A,B>> seqP(Log<A> m1, Log<B> m2) {
     // below is a first-order syntactic sugar
     m1.flatMap( x ->
	   m2.map (y -> new Pair(x,y))
  }  
```

**Comment by tarshalim** [comment]
I think if the first map transforms each element into another collection or stream, then yep you gotta use flatMap afterwards. But if the first map operation transforms each element into a single element, then you can use map again

**Comment by nicklass8** [comment]
hi it is not wrong to use 2 maps consecutively (e.g. Optional.<Integer>of(3).map(x -> x+1).map(x -> x*2)) 
you only need to use flatMap if the function that you want to apply to your value maps the value to another optional. (e.g. Optional.<Integer>of(3).flatMap(x -> Optional.<Integer>of(x))). In this case, flatMap will flatten the Optionals such that your output is only wrapped in one Optional. However if you use map instead, what map does is it takes the value in the original Optional, performs the function, and then wrap it back into another Optional causing your output to be Optional[Optional[3]]. Hope this helps!
